src/vm/analysis/type_checker/tests/mod.rs

```diff
@@ -294,6 +351,138 @@ fn test_lists() {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+}
+
+#[test]
+fn test_native_asserts_max_len() {
+    let good = [
+        "(asserts-max-len! (list 1 2 3 4) u5)"];
+    let expected = ["(optional (list 5 int))"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+}
+
+#[test]
+fn test_buff_asserts_max_len() {
+    let tests = [
+        "(asserts-max-len! \"12345\" u5)",
+        "(asserts-max-len! \"12345\" u8)",
+        "(asserts-max-len! \"12345\" u4)"];
+    let expected = [
+        "(optional (buff 5))",
+        "(optional (buff 8))",
+        "(optional (buff 4))"];
+
+    for (test, expected) in tests.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&test).unwrap()));
+    }
+}
+
+#[test]
+fn test_native_append() {
+    let good = [
+        "(append (list 2 3) 4)",
+        "(append (list u0) u0)"];
+    let expected = ["(list 3 int)", "(list 2 uint)"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+
+    let bad = [
+        "(append (list 2 3) u4)",
+        "(append (list u0) 1)",
+        "(append (list u0))"];
+
+    let bad_expected = [
+        CheckErrors::TypeError(IntType, UIntType),
+        CheckErrors::TypeError(UIntType, IntType),
+        CheckErrors::IncorrectArgumentCount(2, 1),
+    ];
+    for (bad_test, expected) in bad.iter().zip(bad_expected.iter()) {
+        assert_eq!(expected, &type_check_helper(&bad_test).unwrap_err().err);
+    }
+}
+
+#[test]
+fn test_native_concat() {
+    let good = [
+        "(concat (list 2 3) (list 4 5))"];
+    let expected = ["(list 4 int)"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+
+    let bad = [
+        "(concat (list 2 3) (list u4))",
+        "(concat (list u0) (list 1))",
+        "(concat (list u0))"];
+
+    let bad_expected = [
+        CheckErrors::TypeError(IntType, UIntType),
+        CheckErrors::TypeError(UIntType, IntType),
+        CheckErrors::IncorrectArgumentCount(2, 1),
+    ];
+    for (bad_test, expected) in bad.iter().zip(bad_expected.iter()) {
+        assert_eq!(expected, &type_check_helper(&bad_test).unwrap_err().err);
+    }
+}
+
+#[test]
+fn test_buff_concat() {
+    let good = [
+        "(concat \"123\" \"58\")"];
+    let expected = ["(buff 5)"];
+
+    for (good_test, expected) in good.iter().zip(expected.iter()) {
+        assert_eq!(expected, &format!("{}", type_check_helper(&good_test).unwrap()));
+    }
+}
+
+#[test]
+fn test_buff_filter() {
+    let good = [
+        "(define-private (f (e (buff 1))) (eq? e \"1\"))
+        (filter f \"101010\")"];
+    let expected = ["(buff 6)"];
```

Comment:
Wouldn't the type of `(filter f "101010")` be `(list 6 bool)`?

Reply:
`f` is a predicate being used by `filter`. if `f` is true for a given element `e`, then `e` is being returned (and excluded when `f` is false).

Reply:
Thanks for clarifying!